pkg/reconciler/messaging/channel/controller.go

```diff
@@ -49,51 +53,85 @@ const (
 type Constructor injection.ControllerConstructor
 
 // NewConstructor creates a constructor to make a Channel controller.
-func NewConstructor(ipm iam.IAMPolicyManager, gcpas *gcpauth.StoreSingleton) Constructor {
+func NewConstructor(drs *dataresidency.StoreSingleton) Constructor {
 	return func(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
-		return newController(ctx, cmw, ipm, gcpas.Store(ctx, cmw))
+		return newController(ctx, cmw, drs.Store(ctx, cmw))
 	}
 }
 
 func newController(
 	ctx context.Context,
 	cmw configmap.Watcher,
-	ipm iam.IAMPolicyManager,
-	gcpas *gcpauth.Store,
+	drs *dataresidency.Store,
 ) *controller.Impl {
 	channelInformer := channelinformer.Get(ctx)
+	bcInformer := brokercellinformer.Get(ctx)
+
+	var client *pubsub.Client
+	// If there is an error, the projectID will be empty. The reconciler will retry
+	// to get the projectID during reconciliation.
+	projectID, err := utils.ProjectIDOrDefault("")
+	if err != nil {
+		logging.FromContext(ctx).Error("Failed to get project ID", zap.Error(err))
+	} else {
+		// Attempt to create a pubsub client for all worker threads to use. If this
+		// fails, pass a nil value to the Reconciler. They will attempt to
+		// create a client on reconcile.
+		if client, err = pubsub.NewClient(ctx, projectID); err != nil {
+			client = nil
+			logging.FromContext(ctx).Error("Failed to create controller-wide Pub/Sub client", zap.Error(err))
+		}
+	}
 
-	topicInformer := topicinformer.Get(ctx)
-	pullSubscriptionInformer := pullsubscriptioninformer.Get(ctx)
-	serviceAccountInformer := serviceaccountinformers.Get(ctx)
+	if client != nil {
+		go func() {
+			<-ctx.Done()
+			client.Close()
+		}()
+	}
 
 	r := &Reconciler{
-		Base:          reconciler.NewBase(ctx, controllerAgentName, cmw),
-		Identity:      identity.NewIdentity(ctx, ipm, gcpas),
-		channelLister: channelInformer.Lister(),
-		topicLister:   topicInformer.Lister(),
+		Reconciler: celltenant.Reconciler{
+			Base:               reconciler.NewBase(ctx, controllerAgentName, cmw),
+			BrokerCellLister:   bcInformer.Lister(),
+			ProjectID:          projectID,
+			PubsubClient:       client,
+			DataresidencyStore: drs,
+		},
+		targetReconciler: &celltenant.TargetReconciler{
+			ProjectID:          projectID,
+			PubsubClient:       client,
+			DataresidencyStore: drs,
+		},
 	}
 	impl := channelreconciler.NewImpl(ctx, r)
 
 	r.Logger.Info("Setting up event handlers")
 	channelInformer.Informer().AddEventHandler(controller.HandleAll(impl.Enqueue))
```

Comment:
Will this add all the channels? In broker there is a filter checking the broker class now.

Reply:
> Should we add a resync period like in the broker controller?

There is already a default resync period for all these informers. Looking through the code, it seems to be 10 hours:

https://github.com/google/knative-gcp/blob/c1fd5094c5d21fab13cd5dc42d3ccf0d2d4233bc/vendor/knative.dev/pkg/controller/controller.go#L51

As far as I can tell, nothing calls `WithResyncPeriod` on the context, so the default should still be used. I can still explicitly add it to match the Broker controller, if you'd like.

> Will this add all the channels? In broker there is a filter checking the broker class now.

Yes, this will add all the Channels. Note that this Channel CRD is unique to knative-gcp because it is in the `messaging.cloud.google.com` API group. So we don't have the same issue we have with Broker, where all Broker implementations use the same CRD and therefore need to resort to checking a label to determine which ones to manage. All copies of this Channel CRD are managed by this controller. Therefore we don't need the extra filtering.

Reply:
I would vote for adding the same resync period as the broker for consistency. One fo the big reason was to recreate topics that are deleted accidentally.

Reply:
Done.

Reply:
Yes, most of our controllers have the 5 minute resync cycle now.